src/cunumeric/index/advanced_indexing.cc

```diff
@@ -0,0 +1,109 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+#include "cunumeric/index/advanced_indexing.h"
+#include "cunumeric/index/advanced_indexing_template.inl"
+
+namespace cunumeric {
+
+using namespace Legion;
+using namespace legate;
+
+template <LegateTypeCode CODE, int DIM1, int DIM2>
+struct AdvancedIndexingImplBody<VariantKind::CPU, CODE, DIM1, DIM2> {
+  using VAL = legate_type_of<CODE>;
+
+  void compute_output(Buffer<VAL>& out,
+                      const AccessorRO<VAL, DIM1>& input,
+                      const AccessorRO<bool, DIM2>& index,
+                      const Pitches<DIM1 - 1>& pitches_input,
+                      const Rect<DIM1>& rect_input,
+                      const Pitches<DIM2 - 1>& pitches_index,
+                      const Rect<DIM2>& rect_index,
+                      const size_t volume) const
+  {
+    int64_t out_idx = 0;
+    for (size_t idx = 0; idx < volume; ++idx) {
+      auto p = pitches_index.unflatten(idx, rect_index.lo);
+      if (index[p] == true) {
+        auto p_input = pitches_input.unflatten(idx, rect_input.lo);
+        out[out_idx] = input[p_input];
+        out_idx++;
+      }
+    }
+  }
+
+  void compute_output(Buffer<Point<DIM1>>& out,
+                      const AccessorRO<VAL, DIM1>&,
+                      const AccessorRO<bool, DIM2>& index,
+                      const Pitches<DIM1 - 1>& pitches_input,
+                      const Rect<DIM1>& rect_input,
+                      const Pitches<DIM2 - 1>& pitches_index,
+                      const Rect<DIM2>& rect_index,
+                      const size_t volume) const
+  {
+    int64_t out_idx = 0;
+    for (size_t idx = 0; idx < volume; ++idx) {
+      auto p = pitches_index.unflatten(idx, rect_index.lo);
+      if (index[p] == true) {
+        auto p_input = pitches_input.unflatten(idx, rect_input.lo);
+        out[out_idx] = p_input;
+        out_idx++;
+      }
+    }
+  }
+
+  template <typename OUT_TYPE>
+  size_t operator()(Buffer<OUT_TYPE>& out,
+                    const AccessorRO<VAL, DIM1>& input,
+                    const AccessorRO<bool, DIM2>& index,
+                    const Pitches<DIM1 - 1>& pitches_input,
+                    const Rect<DIM1>& rect_input,
+                    const Pitches<DIM2 - 1>& pitches_index,
+                    const Rect<DIM2>& rect_index) const
+  {
+#ifdef DEBUG_CUNUMERIC
+    // in this case shapes for input and index arrays  should be the same
+    assert(rect_input == rect_index);
```

Comment:
This line is causing an issue for me when compiling on clang:

```
cunumeric/index/advanced_indexing.cc:80:23: fatal error: invalid operands to binary expression ('const Rect<1>' (aka 'const Rect<1, long long>') and 'const Rect<2>' (aka 'const Rect<2, long long>'))
    assert(rect_input == rect_index);
           ~~~~~~~~~~ ^  ~~~~~~~~~~
/Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/include/assert.h:93:25: note: expanded from macro 'assert'
    (__builtin_expect(!(e), 0) ? __assert_rtn(__func__, __FILE__, __LINE__, #e) : (void)0)
                        ^
./cunumeric/index/advanced_indexing_template.inl:59:16: note: in instantiation of function template specialization 'cunumeric::AdvancedIndexingImplBody<cunumeric::VariantKind::CPU, BOOL_LT, 1, 2>::operator()<Realm::Point<1, long long> >' requested here
        size = AdvancedIndexingImplBody<KIND, CODE, DIM1, DIM2>{}(output_arr_set,
               ^
/Users/mpapadakis/legate.core/install/include/core/utilities/dispatch.h:254:25: note: in instantiation of function template specialization 'cunumeric::AdvancedIndexingImpl<cunumeric::VariantKind::CPU, BOOL_LT, 1>::operator()<2>' requested here
      return f.template operator()<2>(std::forward<Fnargs>(args)...);
                        ^
/Users/mpapadakis/legate.core/install/include/core/utilities/dispatch.h:30:27: note: in instantiation of function template specialization 'cunumeric::AdvancedIndexingHelper<cunumeric::VariantKind::CPU>::operator()<BOOL_LT, 1>' requested here
        return f.template operator()<LegateTypeCode::BOOL_LT, DIM>(std::forward<Fnargs>(args)...);
                          ^
cunumeric/index/advanced_indexing.cc:98:3: note: in instantiation of function template specialization 'cunumeric::advanced_indexing_template<cunumeric::VariantKind::CPU>' requested here
  advanced_indexing_template<VariantKind::CPU>(context);
  ^
```

The following change fixes it, but this check should really only be emitted when `DIM1 == DIM2`:

```
assert(Domain(rect_input) == Domain(rect_index));
```

Similar checks exist in the OpenMP and GPU variants of this task.